src/libPersistence/Retriever.cpp

```diff
@@ -50,26 +47,105 @@ bool Retriever::RetrieveTxBlocks(bool trimIncompletedBlocks) {
 
   unsigned int extra_txblocks = (lastBlockNum + 1) % NUM_FINAL_BLOCK_PER_POW;
 
-  if (trimIncompletedBlocks) {
-    // truncate the extra final blocks at last
-    for (unsigned int i = 0; i < extra_txblocks; ++i) {
-      BlockStorage::GetBlockStorage().DeleteTxBlock(lastBlockNum - i);
-      blocks.pop_back();
-    }
-  }
-
-  for (const auto& block : blocks) {
-    m_mediator.m_node->AddBlock(*block);
-  }
-
   /// Retrieve final block state delta from last DS epoch to
-  /// current TX epoch
+  /// current TX epoch and buffer the statedelta for each.
   for (const auto& block : blocks) {
     if (block->GetHeader().GetBlockNum() >= lastBlockNum + 1 - extra_txblocks) {
       bytes stateDelta;
       BlockStorage::GetBlockStorage().GetStateDelta(
```

Comment:
Just notice that, maybe we should handle the boolean return of this function.

Reply:
Fixed. If any of extra statedelta missing in BlockStorage, simple trim the extra tx blocks. we can then fetch the rest  from lookup.